README.md
---
# py_github_cli

[![CI](https://github.com/sbeaulie/py_github_cli/actions/workflows/main.yml/badge.svg)](https://github.com/sbeaulie/py_github_cli/actions/workflows/main.yml)

py_github_cli created by sbeaulie

## Usage

```bash
$ python -m py_github_cli
#or
$ py_github_cli
Usage:
    py_github_cli releases <repo>
    py_github_cli prs <repo>
    py_github_cli -h|--help
```

## Example usage

![Usage](https://github.com/sbeaulie/py_github_cli/blob/main/gifs/usage.gif)

## Supported commands
There are two cli commands supported: `releases` and `prs`
### releases
`py_github_cli releases` -> to get the top 3 releases names and versions
### prs
`py_github_cli prs` -> to get the top 3 prs

## Supported repos

After the command (releases or prs) the first parameter is the repo name

`py_github_cli releases <repo>`
### Repo format
1. OWNER/REPO for example `monkey/github_cli`
2. `https://github.com/OWNER/REPO` or
3. `git@github.com:OWNER/REPO`
### Public repos
Public repos can be queried by anyone without authentication.
### Private repos
Private repos can be queried only if you provide a user and token for access.
If querying a private repo without them, you will get a 404 error (same as not found)
```bash
$ export USER=foo; export TOKEN=ghp_foobar
```
A personal access token can be created in the github UI
1. Go to Settings/Developer settings (https://github.com/settings/tokens)
2. Click generate new token
3. Select the permission for 'repo'
4. save and take note of the token
5. `export TOKEN=${TOKEN_FROMSTEP_4}` or set it before running the program

## Known limitations
### Github rate limiting
60 requests per hour when unauthenticated

5000 requests per hour when authenticated via the USER and TOKEN environment variables

See https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting

## Development

Read the [CONTRIBUTING.md](CONTRIBUTING.md) file.

## Docker-compose
Perhaps the easiest way to run this code without having to install python and virtualenv
is to run it via `docker-compose` which is set to use an off-the-shelf python alpine 
image (only 170MB)

```bash
# optional export ENV vars see above for private repos
$ docker-compose run cli py_github_cli prs vuejs/vue
Received response: 200 https://api.github.com/repos/vuejs/vue/pulls?per_page=3
Listing top 3 PRs
      
  NUMBER  TITLE                                                    
    
  12738   fix: vue ssr server error when mapped item is undefined  
  12737   feat(types): new Vue() improvements (#12730)             
  12735   chore(v3/lifecycle): Enhanced type for onErrorCaptured   
```

## With python installed locally
You can build and run the program via virtualenv. 
```bash
$ make virtualenv
$ source .venv/bin/activate
$ py_github_cli
Usage:
    py_github_cli releases <repo>
    py_github_cli prs <repo>
    py_github_cli -h|--help
```


HISTORY.md
Changelog
=========


0.1.0 (2022-08-15)
------------------
Initial release of py_github_cli with support for 2 github repository queries:
- list the top 3 releases in a repo
- list the top 3 PRs in a repo


tests/test_cli.py
import pytest
import py_github_cli.cli
from unittest.mock import MagicMock
from unittest.mock import patch


@pytest.mark.parametrize(
    "input_repository_string, expected",
    [
        ("a/b", "a/b"),
        ("foo/bar", "foo/bar"),
        ("https://github.com/foo/bar", "foo/bar"),
        ("http://github.com/foo/bar", "foo/bar"),
        ("http://github.com/foo/bar/", "foo/bar"),
        ("git@github.com:foo/bar/", "foo/bar"),
        ("special/characters123_-.", "special/characters123_-."),
        (
            "https://github.com/kamranahmedse/developer-roadmap.git",
            "kamranahmedse/developer-roadmap",
        ),
    ],
)
def test_cli_validate(input_repository_string, expected):
    print(input_repository_string, expected)
    assert py_github_cli.cli.validate(input_repository_string) == expected


@pytest.mark.parametrize(
    "input_repository_string",
    ["a", "foobar", "inval!d/n@me"],
)
def test_cli_validate_raises_exception(input_repository_string):
    with pytest.raises(ValueError):
        py_github_cli.cli.validate(input_repository_string)


@patch("py_github_cli.github_api.repository.get_releases")
def test_cli_main_releases(mock_api_call):
    py_github_cli.cli.ver = MagicMock(return_value="MockVersion")
    py_github_cli.cli.docopt = MagicMock(
        return_value={"releases": True, "prs": False, "<repo>": "foo/bar"}
    )
    py_github_cli.cli.validate = MagicMock(return_value="foo/bar")
    mock_api_call.return_value = MagicMock()
    py_github_cli.cli.main()
    mock_api_call.assert_called_once()


@patch("py_github_cli.github_api.repository.get_prs")
def test_cli_main_prs(mock_api_call):
    py_github_cli.cli.ver = MagicMock(return_value="MockVersion")
    py_github_cli.cli.docopt = MagicMock(
        return_value={"releases": False, "prs": True, "<repo>": "foo/bar"}
    )
    py_github_cli.cli.validate = MagicMock(return_value="foo/bar")
    mock_api_call.return_value = MagicMock()
    py_github_cli.cli.main()
    mock_api_call.assert_called_once()


tests/test_base.py
from py_github_cli.base import NAME


def test_base():
    assert NAME == "py_github_cli"


tests/test_github_api.py
from py_github_cli.github_api import repository
from unittest.mock import MagicMock
from unittest.mock import patch
import pytest

import json
import requests
from requests import Response
from collections import OrderedDict

releases_json = json.loads(
    """[
    {
        "url": "https://api.github.com/repos/octocat/Hello-World/releases/1",
        "html_url": "https://github.com/octocat/Hello-World/releases/v1.0.0",
        "assets_url": "https://api.github.com/repos/octocat/Hello-World/releases/1/assets",
        "upload_url": "https://uploads.github.com/repos/octocat/Hello-World/releases/1/assets{?name,label}",
        "tarball_url": "https://api.github.com/repos/octocat/Hello-World/tarball/v1.0.0",
        "zipball_url": "https://api.github.com/repos/octocat/Hello-World/zipball/v1.0.0",
        "id": 1,
        "node_id": "MDc6UmVsZWFzZTE=",
        "tag_name": "v1.0.0",
        "target_commitish": "master",
        "name": "v1.0.0",
        "body": "Description of the release",
        "draft": false,
        "prerelease": false,
        "created_at": "2013-02-27T19:35:32Z",
        "published_at": "2013-02-27T19:35:32Z"
    }
    ]"""
)

prs_json = json.loads(
    """[
  {
    "url": "https://api.github.com/repos/octocat/Hello-World/pulls/1347",
    "id": 1,
    "node_id": "MDExOlB1bGxSZXF1ZXN0MQ==",
    "html_url": "https://github.com/octocat/Hello-World/pull/1347",
    "diff_url": "https://github.com/octocat/Hello-World/pull/1347.diff",
    "patch_url": "https://github.com/octocat/Hello-World/pull/1347.patch",
    "issue_url": "https://api.github.com/repos/octocat/Hello-World/issues/1347",
    "commits_url": "https://api.github.com/repos/octocat/Hello-World/pulls/1347/commits",
    "review_comments_url": "https://api.github.com/repos/octocat/Hello-World/pulls/1347/comments",
    "review_comment_url": "https://api.github.com/repos/octocat/Hello-World/pulls/comments{/number}",
    "comments_url": "https://api.github.com/repos/octocat/Hello-World/issues/1347/comments",
    "statuses_url": "https://api.github.com/repos/octocat/Hello-World/statuses/6dcb09b5b57875f334f61aebed695e2e4193db5e",
    "number": 1347,
    "state": "open",
    "locked": true,
    "title": "Amazing new feature",
    "user": {
      "login": "octocat",
      "id": 1,
      "node_id": "MDQ6VXNlcjE=",
      "avatar_url": "https://github.com/images/error/octocat_happy.gif",
      "gravatar_id": "",
      "url": "https://api.github.com/users/octocat",
      "html_url": "https://github.com/octocat",
      "followers_url": "https://api.github.com/users/octocat/followers",
      "following_url": "https://api.github.com/users/octocat/following{/other_user}",
      "gists_url": "https://api.github.com/users/octocat/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/octocat/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/octocat/subscriptions",
      "organizations_url": "https://api.github.com/users/octocat/orgs",
      "repos_url": "https://api.github.com/users/octocat/repos",
      "events_url": "https://api.github.com/users/octocat/events{/privacy}",
      "received_events_url": "https://api.github.com/users/octocat/received_events",
      "type": "User",
      "site_admin": false
    },
    "body": "Please pull these awesome changes in!",
    "active_lock_reason": "too heated",
    "created_at": "2011-01-26T19:01:12Z",
    "updated_at": "2011-01-26T19:01:12Z",
    "closed_at": "2011-01-26T19:01:12Z",
    "merged_at": "2011-01-26T19:01:12Z",
    "merge_commit_sha": "e5bd3914e2e596debea16f433f57875b5b90bcd6",
    "_links": {
      "self": {
        "href": "https://api.github.com/repos/octocat/Hello-World/pulls/1347"
      },
      "html": {
        "href": "https://github.com/octocat/Hello-World/pull/1347"
      },
      "issue": {
        "href": "https://api.github.com/repos/octocat/Hello-World/issues/1347"
      },
      "comments": {
        "href": "https://api.github.com/repos/octocat/Hello-World/issues/1347/comments"
      },
      "review_comments": {
        "href": "https://api.github.com/repos/octocat/Hello-World/pulls/1347/comments"
      },
      "review_comment": {
        "href": "https://api.github.com/repos/octocat/Hello-World/pulls/comments{/number}"
      },
      "commits": {
        "href": "https://api.github.com/repos/octocat/Hello-World/pulls/1347/commits"
      },
      "statuses": {
        "href": "https://api.github.com/repos/octocat/Hello-World/statuses/6dcb09b5b57875f334f61aebed695e2e4193db5e"
      }
    },
    "author_association": "OWNER",
    "auto_merge": null,
    "draft": false
  }
  ]"""
)


@patch("requests.get")
def test_github_api_http_get_request(mock_api_call):
    return_response = MagicMock(status_code=200, url="foobar", spec=Response)
    mock_api_call.return_value = return_response
    mock_api_call.return_value.json.return_value = releases_json
    repo = repository("OWENER/REPO")
    result = repository._http_get_request(repo, "foo")
    assert result == releases_json


@patch("requests.get")
def test_github_api_http_get_request_404(mock_api_call):
    return_response = MagicMock(status_code=404, url="foobar", spec=Response)
    mock_api_call.return_value = return_response
    repo = repository("OWENER/REPO")
    result = repository._http_get_request(repo, "foo")
    assert result == None


@patch("requests.get")
def test_github_api_http_get_request_500(mock_api_call):
    return_response = MagicMock(status_code=500, url="foobar", spec=Response)
    mock_api_call.return_value = return_response
    repo = repository("OWENER/REPO")
    result = repository._http_get_request(repo, "foo")
    assert result == None


@patch("requests.get")
def test_github_api_http_get_request_connectionfailure(mock_api_call):
    return_response = MagicMock(side_effect=requests.exceptions.Timeout())
    mock_api_call.side_effect = return_response
    repo = repository("OWENER/REPO")
    with pytest.raises(ConnectionError):
        result = repository._http_get_request(repo, "foo")


def test_github_api_get_releases():
    repository._http_get_request = MagicMock(return_value=releases_json)
    repo = repository("OWENER/REPO")
    assert repository.get_releases(repo) == [["v1.0.0", "v1.0.0"]]


def test_github_api_get_releases_empty():
    repository._http_get_request = MagicMock(return_value=json.loads("[]"))
    repo = repository("OWENER/REPO")
    assert repository.get_releases(repo) == []


def test_github_api_get_prs():
    repository._http_get_request = MagicMock(return_value=prs_json)
    repo = repository("OWENER/REPO")
    assert repository.get_prs(repo) == [[1347, "Amazing new feature"]]


def test_github_api_get_prs_empty():
    repository._http_get_request = MagicMock(return_value=json.loads("[]"))
    repo = repository("OWENER/REPO")
    assert repository.get_prs(repo) == []


py_github_cli/cli.py
"""CLI interface for py_github_cli project.

Usage:
    py_github_cli releases <repo>
    py_github_cli prs <repo>
    py_github_cli -h|--help

Options:
    -h --help       Show this screen.
"""
import regex
from docopt import docopt

from .github_api import repository


def main():  # pragma: no cover
    """
    The main function executes on commands:
    `python -m py_github_cli` and `$ py_github_cli `.
    """
    arguments = docopt(__doc__)
    repo = validate(arguments["<repo>"])

    github = repository(repo)
    if arguments["releases"]:
        github.get_releases()

    if arguments["prs"]:
        github.get_prs()


def validate(document):
    """Method to validate and sanitize the repository input

    Args:
        document (str): The github repository to validate

    Raises:
        ValueError: Raises and displays an error when the repository name is
        not valid

    Returns:
        str: Returns the sanitized github repo in the format OWNER/REPO
    """
    ok = regex.match(
        r"(https?://|git@)?(github.com(/|:))?([\w.-]+/[\w.-]+)", document
    )
    if not ok:
        raise ValueError(
            "Please enter a valid github repository"
            + " in the format myorganization/myrepo"
        )
    # get only the OWNER/REPO part of the match,
    # and remove any trailing '.git' string
    return regex.sub(r"\.git$", "", ok.groups()[-1])


.github/rename_project.sh
#!/usr/bin/env bash
while getopts a:n:u:d: flag
do
    case "${flag}" in
        a) author=${OPTARG};;
        n) name=${OPTARG};;
        u) urlname=${OPTARG};;
        d) description=${OPTARG};;
    esac
done

echo "Author: $author";
echo "Project Name: $name";
echo "Project URL name: $urlname";
echo "Description: $description";

echo "Renaming project..."

original_author="sbeaulie"
original_name="py_github_cli"
original_urlname="py_github_cli"
original_description="Awesome py_github_cli created by sbeaulie"
# for filename in $(find . -name "*.*") 
for filename in $(git ls-files) 
do
    sed -i "s/$original_author/$author/g" $filename
    sed -i "s/$original_name/$name/g" $filename
    sed -i "s/$original_urlname/$urlname/g" $filename
    sed -i "s/$original_description/$description/g" $filename
    echo "Renamed $filename"
done

mv py_github_cli $name

# This command runs only once on GHA!
rm -rf .github/template.yml


py_github_cli/github_api.py
"""
py_github_cli github_api module.
"""
import os

import requests
from columnar import columnar
from requests.auth import HTTPBasicAuth


class repository:
    """The class is to be used for all github API methods that act on a
    repository, for example the API endpoints
    OWNER/REPO/releases
    OWNER/REPO/prs
    This class should be extended to support other GET API endpoints and other
    verbs.
    GET API methods:
    get_releases
    get_prs

    Raises:
        ConnectionError: Raises and displays an error when the connection to
        github could not be established
    """

    PREFIX = "https://api.github.com/repos/"

    def __init__(self, repo):
        """The class constructor saves important data

        Args:
            repo (str): The sanitized input repository in the format OWNER/REPO
        """
        self.repo = repo
        self.token = os.environ.get("TOKEN")
        self.user = os.environ.get("USER")

    def get_releases(self):
        """Main method the the command py_github_api releases <repo>
        Prints the top 3 releases when available

        Returns:
            list: A list containing the 3 records
        """
        headers = ["name", "tag name"]
        lines = []
        json_response = self._http_get_request("releases")
        if json_response and len(json_response) > 0:
            print("Listing top 3 releases")
            for r in json_response:
                lines.append([r["name"], r["tag_name"]])
            table = columnar(lines, headers, no_borders=True)
            print(table)
        elif json_response is not None:
            print("No release yet")
        return lines

    def get_prs(self):
        """Main method the the command py_github_api prs <repo>
        Prints the top 3 PRs when available

        Returns:
            list: A list containing the 3 records
        """
        headers = ["number", "title"]
        lines = []
        json_response = self._http_get_request("pulls")
        if json_response and len(json_response) > 0:
            print("Listing top 3 PRs")
            for r in json_response:
                lines.append([r["number"], r["title"]])
            table = columnar(lines, headers, no_borders=True)
            print(table)
        elif json_response is not None:
            print("No prs yet")
        return lines

    def _http_get_request(self, query):
        """A utility method to query the github API with a HTTP GET request
        This method is generic enough to query any API endpoint
        It displays the minimal information from the response such as the
        status_code and the URL that was queried.

        Args:
            query (str): the API to query with a HTTP GET request, for example
            'releases'

        Raises:
            ConnectionError: Raises and displays an error when the connection
            to github could not be established

        Returns:
            json: The HTTP body response in json format or None
        """
        try:
            uri = self.PREFIX + self.repo + "/" + query
            response = requests.get(
                uri,
                params={"per_page": "3"},
                headers={"Accept": "application/vnd.github+json"},
                auth=HTTPBasicAuth(self.user, self.token),
            )
            print("Received response:", response.status_code, response.url)
            if response.status_code == 200:
                json_response = response.json()
                return json_response
            elif response.status_code == 404:
                print(
                    "Received a 404 error, which means either the repo does",
                    "not exist, or it is a private repo. In order to query",
                    "private repos, please set the ENV var USER and TOKEN",
                    "with a personal access token. See README.md for details.",
                )
            else:
                print("Invalid response received from github")
        except (
            requests.exceptions.ConnectionError,
            requests.exceptions.Timeout,
            requests.exceptions.RequestException,
        ) as ex:
            raise ConnectionError(ex)
